### PR TITLE
Improvements to existing rules and 3 new rules

### PR DESCRIPTION
**Pull request by GotenJBZ**
This PR introduces or improve several security-focused Semgrep rules for Move smart contracts on Aptos:

- constructor-ref-leak: Detects leaking of ConstructorRef via public functions
- missing-ownership-check: Identifies functions missing object ownership verification
- public-friend-entry: Catches unintended exposure from combined friend/entry functions
- public-randomness: Detects unsafe public access to randomness functions
- signer-leak: Identifies dangerous exposure of signer capabilities
- objects-in-group-stored-without-account: Checks for unsafe object transfers within resource groups
- object-without-signer: Detects object operations without signer verification

All rules include test cases, documentation links

### CODE REVIEW

**Comment by gedigi** [comment]
can you add a meaningful description about what this PR addresses?

File: rules/move_on_aptos/objects/security/missing-ownership-check.yaml [hunk 1]
```diff
@@ -17,58 +18,220 @@ rules:
...
+          - pattern: |
+              aptos_framework::object::transfer<$TYPE>($SIGNER, $OBJ, $ANY_VAR);
+          - pattern: |
+              aptos_framework::object::transfer_to_object<$TYPE>($SIGNER,$OBJ, $ANY_VAR);
+
+            #start test
+          - pattern-inside: |
+              aptos_framework::object::is_owner<$TYPE>($OBJ, aptos_framework::signer::address_of($SIGNER))
+        
+          - pattern-inside: |
```
Comment (zi0Black) [inline rules/move_on_aptos/objects/security/missing-ownership-check.yaml:84]: Why do we have some `pattern` and some `pattern-inside` matching a similar pattern? Can't we just use `pattern`?

File: rules/move_on_aptos/lang/security/signer-leak.yaml [hunk 2]
```diff
@@ -40,13 +47,4 @@ rules:
         - pattern: |
             public fun $FUN(...): (..., signer, ...)
         - pattern: |
-            public fun $FUN(...): (..., &signer, ...)
-      #
```
Comment (zi0Black) [inline rules/move_on_aptos/lang/security/signer-leak.yaml:44]: Why this comment has been removed?

File: rules/move_on_aptos/lang/security/signer-leak.yaml [hunk 3]
```diff
@@ -40,13 +47,4 @@ rules:
...
-      # Metavariable matching reference is broken on the Semgrep side. For now, we
-      # will manually enumerate the patterns (`signer` and `&signer`).
-      #
-      # - metavariable-pattern:
-      #     metavariable: $RETURN
-      #     pattern-either:
-      #       - pattern: signer
-      #       - pattern: "&signer"
```
Comment (zi0Black) [inline rules/move_on_aptos/lang/security/signer-leak.yaml:52]: Why this comment has been removed?

File: rules/move_on_aptos/objects/security/missing-ownership-check.yaml (hunk 1 above)
Comment (GotenJBZ) [reply rules/move_on_aptos/objects/security/missing-ownership-check.yaml:84]: Matching `aptos_framework::object::is_owner` is actually broken. I was trying different patterns, but no one worked as expected

File: rules/move_on_aptos/lang/security/signer-leak.yaml (hunk 3 above)
Comment (GotenJBZ) [reply rules/move_on_aptos/lang/security/signer-leak.yaml:52]: Clean up. The rule works perfectly right now. No need to use `metavariable-pattern`

Comment (zi0Black) [reply rules/move_on_aptos/lang/security/signer-leak.yaml:52]: >The rule works perfectly right now. No need to use metavariable-pattern.
The comment was suggesting using a more concise way of writing the patterns. Either verify that this is fixed and use metavariables, or maintain this comment.

**Review by zi0Black — APPROVED** [review]

File: rules/move_on_aptos/lang/security/public-friend-entry.yaml [hunk 4]
```diff
@@ -1,29 +1,36 @@
...
-    message: Using both the scope identifier `friend` and `entry` in a function make it callable by any transaction, even if public(friend) is used. This can inadvertently expose sensitive functionality to unauthorized users, potentially leading to security vulnerabilities. 
+    message: Using both the scope identifier `friend` and `entry` in a function make it callable by any transaction, even if public(friend) is used. This can inadvertently expose sensitive functionality to unauthorized users.
     severity: WARNING
     metadata:
-      likelihood: HIGH
-      impact: HIGH
-      confidence: HIGH
+      likelihood: MEDIUM
+      impact: MEDIUM
+      confidence: MEDIUM
```
[the remark this answers is not in the document]
Comment (GotenJBZ) [reply rules/move_on_aptos/lang/security/public-friend-entry.yaml:9]: No, but having `public(friend) entry` is not a vulnerability in itself, there are many cases where it has no impact at all

File: rules/move_on_aptos/objects/security/object-without-signer.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+rules:
+  - id: object-without-signer
```
[the remark this answers is not in the document]
Comment (GotenJBZ) [reply rules/move_on_aptos/objects/security/object-without-signer.yaml:2]: Yes, it is. But if there is no signer the check is easier. So I decided to split this it

**Pull request merged**